### PR TITLE
Core 4163

### PR DESCRIPTION
**Pull request by abdyfranco**

### CODE REVIEW

File: pterodactyl.php [hunk 1]
```diff
@@ -849,6 +851,31 @@ private function actionsTab($package, $service, $client = false, array $get = nu
...
+     * Returns an array of package fields, overriding the configurable options
+     *
+     * @param array $vars An array of key/value input pairs
+     * @param stdClass $package A stdClass object representing the package for the service
+     * @return stdClass The modified package object
+     */
+    private function getConfigurableOptions(array $vars, $package)
+    {
+        $fields = [
+            'location_id', 'egg_id', 'memory', 'cpu', 'disk', 'io'
```
Comment (JReissmueller) [inline pterodactyl.php]: What about nest, port range, memory, etc?

File: pterodactyl.php [hunk 2]
```diff
@@ -239,6 +238,9 @@ public function addService(
             return;
         }
 
+        // Set configurable options
+        $package = $this->getConfigurableOptions($vars, $package);
```
Comment (JReissmueller) [inline pterodactyl.php:242]: What about edit? Or will that be included in CORE-4123?

**Review by JReissmueller — CHANGES_REQUESTED** [review]

File: pterodactyl.php (hunk 2 above)
Comment (abdyfranco) [reply pterodactyl.php:242]: Yes, to be able to modify the configurable options it is necessary that the package can be modified, so it seems that it is more suitable for CORE-4123.

**Comment by JReissmueller** [comment]
I think we'll need to do this a little more thoroughly.  All the memory and such work well, but nest/egg/location will require the service fields to be re-fetched.  We should be able to use the InputFields::setHtml() method to set some js to refetch options if one of these config options is changed.

**Comment by abdyfranco** [comment]
> I think we'll need to do this a little more thoroughly. All the memory and such work well, but nest/egg/location will require the service fields to be re-fetched. We should be able to use the InputFields::setHtml() method to set some js to refetch options if one of these config options is changed.

Updated

**Review by JReissmueller — APPROVED** [review]

File: lib/pterodactyl_service.php [hunk 3]
```diff
@@ -329,6 +329,21 @@ public function getFields($pterodactylEgg, $package, $vars = null, $admin = fals
...
+        $fields->setHtml("
+            <script type=\"text/javascript\">
+                $(document).ready(function() {
+                    // Re-fetch module options to pull in eggs and egg variables
+                    // when a nest or egg respectively is selected
+                    $('.package_options').on('focusout', '[name^=\"configoptions\"]', function() {
+                        var form = $(this).closest('form');
+                        $(form).append('<input type=\"hidden\" name=\"refresh_fields\" value=\"true\">');
+                        $(form).submit();
+                    );
```
Comment (JReissmueller) [inline lib/pterodactyl_service.php:351]: Function needs a closing bracket

**Review by JReissmueller — CHANGES_REQUESTED** [review]
getAdminAddFields() and getClientAddFields() need to fetch service fields based on config options if set.

File: pterodactyl.php [hunk 4]
```diff
@@ -1208,11 +1240,36 @@ public function getAdminAddFields($package, $vars = null)
...
+        }
+        if (!isset($this->Form)) {
+            Loader::loadHelpers($this, ['Form']);
+        }
+
+        $configurable_options = $this->Form->collapseObjectArray(
+            $this->Record->select(['package_options.id', 'package_options.name'])
+                ->from('package_options')
+                ->where('package_options.company_id', '=', Configure::get('Blesta.company_id'))
+                ->where('package_options.id', 'IN', array_keys($vars->configoptions))
```
Comment (JReissmueller) [inline pterodactyl.php]: I get this error when visiting the admin service add page:
Undefined property: stdClass::$configoptions on line 1255 in ...\blesta-5.0\public_html\components\modules\pterodactyl\pterodactyl.php

File: lib/pterodactyl_service.php [hunk 5]
```diff
@@ -329,6 +329,21 @@ public function getFields($pterodactylEgg, $package, $vars = null, $admin = fals
             }
         }
 
+        // Set js to refetch options when the nest or egg is changed
+        $fields->setHtml("
+            <script type=\"text/javascript\">
+                $(document).ready(function() {
+                    // Re-fetch module options to pull in eggs and egg variables
+                    // when a nest or egg respectively is selected
+                    $('.package_options').on('focusout', '[name^=\"configoptions\"]', function() {
```
Comment (JReissmueller) [inline lib/pterodactyl_service.php]: This reloads whenever any config option is changed.  If possible we'd like to only reload when the location, nest, or egg is changed

**Review by JReissmueller — CHANGES_REQUESTED** [review]
I left a couple comments above

File: pterodactyl.php (hunk 4 above)
Comment (JReissmueller) [reply pterodactyl.php]: I updated this part

File: lib/pterodactyl_service.php (hunk 5 above)
Comment (JReissmueller) [reply lib/pterodactyl_service.php]: I've updated this, but could use a second pair of eyes for testing.

File: pterodactyl.php [hunk 6]
```diff
@@ -1323,23 +1323,41 @@ public function getClientAddFields($package, $vars = null)
...
-        // Load egg
-        $nest_id = isset($configurable_options['nest_id'], $vars->configoptions[$configurable_options['nest_id']])
-            ? $vars->configoptions[$configurable_options['nest_id']]
-            : $package->meta->nest_id;
-        $egg_id = isset($configurable_options['egg_id'], $vars->configoptions[$configurable_options['egg_id']])
-            ? $vars->configoptions[$configurable_options['egg_id']]
-            : $package->meta->egg_id;
+        // Load nest/egg from config option if submitted
+        if (isset($vars->configoptions)) {
+            if (isset($configurable_options['nest_id'], $vars->configoptions[$configurable_options['nest_id']])) {
```
Comment (abdyfranco) [inline pterodactyl.php]: $configurable_options seems to be undefined here

File: pterodactyl.php [hunk 7]
```diff
@@ -1251,18 +1251,30 @@ public function getAdminAddFields($package, $vars = null)
...
-                $this->Record->select(['package_options.id', 'package_options.name'])
-                    ->from('package_options')
-                    ->where('package_options.company_id', '=', Configure::get('Blesta.company_id'))
-                    ->where('package_options.id', 'IN', array_keys($vars->configoptions))
-                    ->fetchAll(),
-                'id',
-                'name'
-            );
-
             if (isset($configurable_options['nest_id'], $vars->configoptions[$configurable_options['nest_id']])) {
```
Comment (abdyfranco) [inline pterodactyl.php]: $configurable_options seems to be undefined, maybe you mean $package->configurable_options here?

Comment (JReissmueller) [reply pterodactyl.php]: Updated

**Review by JReissmueller — APPROVED** [review]

**Pull request merged**